### ISSUE TITLE
parse-json escaping rules (fn/parse-json.xml fn-parse-json-028)

### ISSUE THREAD

**Issue by faassen** [w3c/qt3tests#65]
The specification says:

https://www.w3.org/TR/xpath-functions-31/#func-parse-json

> The effect of the one-argument form of this function is the same as calling the two-argument form with an empty map as the value of the $options argument. 

And the `escape` parameter says `Default: true`

I read this as saying that if `parse-json` is called without a map as second parameter JSON escaping is to take place.

But `parse-json-028` to `parse-json-031` all seem to assume that `escape` is `false`:

```
      <test>parse-json('["\\"]')</test>
      <result>
         <all-of><assert>$result?1 = "\"</assert>
         <assert>$result?1 instance of xs:string</assert></all-of>
      </result>
```

`\\` in JSON is `\`. If escape is set to true, then JSON escape sequences should appear in the result. There are tests for this when it's set explicitly to `true`, such as `fn-parse-json-035`.

The tests assume either a different default for `escape` than the specification mandates, or the default is different for the single parameter form of `parse-json`.

P.S. The specification assumes rather particular things about JSON handling. The Rust `json` crate is a reasonable implementation of the spec, but it makes `liberal` meaningless, makes `duplicates` impossible to implement correctly (`use-last` is always in use, which the JSON RFC says is a common compatible interpretation), and JSON escaping rules are handled so that by the time I get a string, I need to re-escape it again in order to follow this specification. It also mystifies me why escaped JSON characters in the item result should ever be desirable.

**Comment by michaelhkay** [comment]
You're right. I see that my implementation, which passes these tests, is using escape="false" as the default. I'm not sure what to do about it: I'll raise it with the QT4 community group, but we can't do anything about XPath 3.1.

**Comment by michaelhkay** [comment]
I raised the issue at https://github.com/qt4cg/qtspecs/issues/1555 -- please feel welcome to join in the discussion.

**Comment by michaelhkay** [comment]
For what it's worth, an earlier draft of XPath 3.1 at

https://www.w3.org/TR/2014/WD-xpath-functions-31-20141007/#func-parse-json

used the options "unescape=true|false" with the default being true, and `unescape=true` corresponded roughly to the behavior of `escape=false` in the final specification. I suspect that there was an oversight when the option was renamed, with a failure to change the default value.

**Comment by faassen** [comment]
The situation with the tests is even stranger. fn-parse-json-061 and fn-parse-json-064 which both get an explicit map (but without explicit `escape`) *do* rely on escape to be true by default, while the tests I reported before that don't get an explicit map rely on escape being false by default.

**Comment by michaelhkay** [comment]
I think the examples given in the spec, especially:

> The expression parse-json('{"x":"\\", "y":"\u0025"}') returns map{"x":"\","y":"%"}.
> 
> The expression parse-json('{"x":"\\", "y":"\u0025"}', map{'escape':true()}) returns map{"x":"\\","y":"%"}.

make it pretty clear that the intended default was escape=false, and that giving the default as true is a simple error.
